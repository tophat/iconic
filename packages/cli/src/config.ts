import { cosmiconfig } from 'cosmiconfig'

import { Configuration } from '@react-vector-graphics/types'

const cache = process.env.NODE_ENV !== 'test'
const explorer = cosmiconfig('rvg', { cache })

export const init = (): Configuration => ({
    options: {},
    plugins: [],
})
export const find = explorer.search
export const load = explorer.load
