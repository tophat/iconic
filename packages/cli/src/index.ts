import * as path from 'path'

import yargs from 'yargs'
import * as fs from 'fs-extra'
import convert from '@react-vector-graphics/core'

function run(src: string, dest: string): void {
    const files = fs.readdirSync(src)
    const filesContents = files.map((file): [string, string] => [
        file,
        fs.readFileSync(path.join(src, file), { encoding: 'utf-8' }),
    ])
    const results = filesContents.map(([file, contents]): object =>
        convert(file, contents),
    )
    for (const result of results) {
        for (const [file, contents] of Object.entries(result)) {
            const filePath = path.join(dest, file)
            fs.outputFileSync(filePath, contents)
        }
    }
}

const argv = yargs
    .usage('Usage: $0 -s [src] -d [dest]')
    .option('src', {
        alias: 's',
        describe: 'source folder',
        type: 'string',
    })
    .option('dest', {
        alias: 'd',
        describe: 'destination folder',
        type: 'string',
    })
    .demandOption(['s', 'd']).argv

const src = argv.s as string
const dest = argv.d as string
run(src, dest)
