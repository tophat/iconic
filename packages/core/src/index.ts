export { getLogger } from './logging'
export { run as default } from './main'
