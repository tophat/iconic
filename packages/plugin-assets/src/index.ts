import * as path from 'path'

import * as fs from 'fs-extra'
import * as glob from 'glob'

import {
    NamingScheme,
    Plugin,
    PluginParams,
} from '@react-vector-graphics/types'

import { pathToName } from './utils'

const ns = (s: TemplateStringsArray): string => `assets/${s}`

export const OPTIONS = {
    FILE_EXT: ns`fileExt`,
    GLOB_PATTERN: ns`globPattern`,
    NAME_SCHEME: ns`nameScheme`,
    OUTPUT_PATH: ns`outputPath`,
}

const run: Plugin = async (code, config, state) => {
    if (code) {
        const outputPath = config.options[OPTIONS.OUTPUT_PATH]
        if (outputPath && state.componentName) {
            const pathToFile = path.join(outputPath, state.componentName)
            const fileExt = config.options[OPTIONS.FILE_EXT]
            const filePath = fileExt ? `${pathToFile}.${fileExt}` : pathToFile
            fs.outputFileSync(filePath, code)
        }
        return { code, state }
    } else {
        const globPattern: string = config.options[OPTIONS.GLOB_PATTERN]
        const nameScheme: NamingScheme = config.options[OPTIONS.NAME_SCHEME]
        return glob.sync(globPattern).map(
            (filePath: string): PluginParams => ({
                code: fs.readFileSync(filePath, { encoding: 'utf-8' }),
                state: Object.assign(
                    {
                        componentName: pathToName(filePath, nameScheme),
                        filePath,
                    },
                    state,
                ),
            }),
        )
    }
}

export default run
