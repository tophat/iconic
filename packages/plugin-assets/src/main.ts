import * as path from 'path'

import * as fs from 'fs-extra'
import * as glob from 'glob'

import {
    Logger,
    NamingScheme,
    Plugin,
    PluginParams,
    State,
} from '@react-vector-graphics/types'

import { pathToName } from './utils'
import { STATE } from './constants'

import { OPTIONS } from '.'

const findAssets = (args: {
    globPattern: string
    nameScheme: NamingScheme
    state: State
}): PluginParams[] => {
    return glob.sync(args.globPattern).map(
        (filePath: string): PluginParams => ({
            code: fs.readFileSync(filePath, { encoding: 'utf-8' }),
            state: Object.assign(
                {
                    [STATE.COMPONENT_NAME]: pathToName(
                        filePath,
                        args.nameScheme,
                    ),
                    [STATE.FILE_PATH]: filePath,
                },
                args.state,
            ),
        }),
    )
}

const writeComponent = (args: {
    assetFile?: string
    code: string
    componentName?: string
    componentFiles: { [fileName: string]: string }
    outputPath?: string
    fileExt?: string
    logger: Logger
}): void => {
    if (!args.componentName) {
        return args.logger.warn(
            `No '${STATE.COMPONENT_NAME}' provided for '${args.assetFile}'.`,
        )
    }
    if (!args.outputPath) {
        return args.logger.warn(`No '${OPTIONS.OUTPUT_PATH}' provided.`)
    }
    if (!args.fileExt) {
        args.logger.warn(`No '${OPTIONS.FILE_EXT}' provided.`)
    }

    const componentFiles = Object.entries(args.componentFiles)
    const pathToFolder = path.join(
        args.outputPath,
        componentFiles.length ? args.componentName : '',
    )
    const pathToFile = path.join(
        pathToFolder,
        componentFiles.length ? 'index' : args.componentName,
    )
    const componentFilePath = args.fileExt
        ? `${pathToFile}.${args.fileExt}`
        : pathToFile
    fs.outputFileSync(componentFilePath, args.code)
    for (const [fileName, fileContents] of componentFiles) {
        const filePath = path.join(pathToFolder, fileName)
        fs.outputFileSync(filePath, fileContents)
    }
}

export const run: Plugin = async (code, config, state, logger) => {
    if (code) {
        writeComponent({
            assetFile: state[STATE.FILE_PATH],
            code,
            componentFiles: state[STATE.COMPONENT_FILES] ?? {},
            componentName: state[STATE.COMPONENT_NAME],
            fileExt: config.options[OPTIONS.FILE_EXT],
            logger,
            outputPath: config.options[OPTIONS.OUTPUT_PATH],
        })
        return { code, state }
    } else {
        return findAssets({
            globPattern: config.options[OPTIONS.GLOB_PATTERN],
            nameScheme: config.options[OPTIONS.NAME_SCHEME],
            state,
        })
    }
}
