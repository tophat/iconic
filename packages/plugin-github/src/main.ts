import { Plugin } from '@react-vector-graphics/types'

import { OPTIONS, STATE } from './constants'
import findAssets from './findAssets'
import writeComponent from './writeComponent'

export const run: Plugin = async (code, config, state, logger) => {
    const github = {
        api: config.options[OPTIONS.API],
        base: config.options[OPTIONS.BASE],
        head: config.options[OPTIONS.HEAD],
        owner: config.options[OPTIONS.OWNER],
        repo: config.options[OPTIONS.REPO],
    }
    const sharedArgs = {
        folderPath: config.options[OPTIONS.FOLDER_PATH],
        github,
    }
    if (code) {
        await writeComponent({
            ...sharedArgs,
            assetFile: state[STATE.FILE_PATH],
            code,
            commitMessagePatterns: {
                create: config.options[OPTIONS.COMMIT_CREATE_PATTERN],
                delete: config.options[OPTIONS.COMMIT_DELETE_PATTERN],
                update: config.options[OPTIONS.COMMIT_UPDATE_PATTERN],
            },
            componentFiles: state[STATE.COMPONENT_FILES] ?? {},
            componentName: state[STATE.COMPONENT_NAME],
            componentNameOld: state[STATE.COMPONENT_NAME_OLD],
            diffType: state[STATE.DIFF_TYPE],
            fileExt: config.options[OPTIONS.FILE_EXT],
            logger,
            outputPath: config.options[OPTIONS.OUTPUT_PATH],
        })
        return { code, state }
    } else {
        return findAssets({
            ...sharedArgs,
            globPattern: config.options[OPTIONS.GLOB_PATTERN],
            nameScheme: config.options[OPTIONS.NAME_SCHEME],
            state,
        })
    }
}
