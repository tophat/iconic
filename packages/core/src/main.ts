import { loadConfig } from '@svgr/core'

import {
    Configuration,
    Logger,
    PluginParams,
} from '@react-vector-graphics/types'

import { getPlugins, resolvePlugin } from './plugins'
import { getLogger } from './logging'

const normalizePluginParams = (
    codeOrParams: string | PluginParams,
    prevParams: Partial<PluginParams>,
): PluginParams => {
    if (typeof codeOrParams === 'string') {
        return { state: {}, ...prevParams, code: codeOrParams }
    }
    return { ...prevParams, ...codeOrParams }
}

export const run = async ({
    config,
    logger = getLogger(),
}: {
    config: Configuration
    logger?: Logger
}): Promise<void> => {
    logger.info('options:', config.options)
    const pluginParams: PluginParams[] = [{} as PluginParams]
    for (const plugin of getPlugins(config)) {
        logger.info('plugin:', plugin)
        const [pluginFn, pluginConfig] = await Promise.all([
            resolvePlugin(plugin),
            loadConfig(config),
        ])
        const results = await Promise.all(
            pluginParams.splice(0).map(async params => {
                const result = await pluginFn(
                    params.code,
                    pluginConfig,
                    params.state,
                    logger,
                )
                return (Array.isArray(result) ? result : [result])
                    .filter(Boolean)
                    .map(r => normalizePluginParams(r, params))
            }),
        )
        pluginParams.push(...pluginParams.concat(...results))
    }
}
