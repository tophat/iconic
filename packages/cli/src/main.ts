import yargs from 'yargs'

import { Configuration } from '@react-vector-graphics/types'
import rvgCore from '@react-vector-graphics/core'
import { OPTIONS } from '@react-vector-graphics/plugin-assets'

const mergeCliOptions = (config: Configuration): Configuration => {
    if (!config.options) config.options = {}
    const { argv } = yargs(process.argv)
        .usage('Usage: $0 -p [pattern] -o [output]')
        .option('pattern', {
            alias: 'p',
            default: config.options[OPTIONS.GLOB_PATTERN],
            describe: 'SVG files glob pattern',
            type: 'string',
        })
        .option('output', {
            alias: 'o',
            default: config.options[OPTIONS.OUTPUT_PATH],
            describe: 'Destination folder',
            type: 'string',
        })
        .option('ext', {
            alias: 'e',
            default: config.options[OPTIONS.FILE_EXT],
            describe: 'Component file extension',
            type: 'string',
        })
        .demandOption(['pattern', 'output'])

    config.options[OPTIONS.GLOB_PATTERN] = argv.pattern
    config.options[OPTIONS.OUTPUT_PATH] = argv.output
    config.options[OPTIONS.FILE_EXT] = argv.ext
    return config
}

export const run = async (config: Configuration): Promise<void> => {
    const assetPlugin = '@react-vector-graphics/plugin-assets'
    if (!config.plugins?.length) {
        config.plugins = [assetPlugin, '@svgr/plugin-jsx', assetPlugin]
    }
    await rvgCore({ config: mergeCliOptions(config) })
}
