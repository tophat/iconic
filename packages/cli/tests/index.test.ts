import * as mockProps from 'jest-mock-props'

import { OPTIONS } from '@react-vector-graphics/plugin-assets'
import rvgCore from '@react-vector-graphics/core'

import { run } from '../src/main'

jest.mock('@react-vector-graphics/core', () => jest.fn())
mockProps.extend(jest)

describe('cli', () => {
    const spyRvgCore = rvgCore as jest.Mocked<typeof rvgCore>
    const spyProcessArgs = jest.spyOnProp(process, 'argv')
    const spyProcessExit = jest.spyOn(process, 'exit')
    const spyLogError = jest.spyOn(console, 'error')

    beforeEach(() => {
        spyProcessExit.mockImplementation(code => {
            throw Error(`Process exited with code: ${code}`)
        })
    })

    afterEach(() => {
        jest.resetAllMocks()
    })

    it('requests for missing option: pattern', async () => {
        const config = {
            options: {
                [OPTIONS.OUTPUT_PATH]: 'tests/component',
            },
            plugins: [],
        }
        await expect(run(config)).rejects.toThrowErrorMatchingSnapshot()
        expect(spyLogError).toHaveBeenCalledWith(
            'Missing required argument: pattern',
        )
    })

    it('requests for missing option: output', async () => {
        const config = {
            options: {
                [OPTIONS.GLOB_PATTERN]: '*.svg',
            },
            plugins: [],
        }
        await expect(run(config)).rejects.toThrowErrorMatchingSnapshot()
        expect(spyLogError).toHaveBeenCalledWith(
            'Missing required argument: output',
        )
    })

    it('runs with combination of supplied options', async () => {
        spyProcessArgs.mockValue([
            'node',
            'test.js',
            '--ext',
            'js',
            '--output',
            'tests/components',
        ])
        const config = {
            options: { [OPTIONS.GLOB_PATTERN]: '*.svg' },
            plugins: [],
        }
        await run(config)
        expect(spyLogError).not.toHaveBeenCalled()
        expect(spyRvgCore).toHaveBeenCalledWith({
            config: {
                options: {
                    [OPTIONS.FILE_EXT]: 'js',
                    [OPTIONS.GLOB_PATTERN]: '*.svg',
                    [OPTIONS.OUTPUT_PATH]: 'tests/components',
                },
                plugins: [
                    '@react-vector-graphics/plugin-assets',
                    '@svgr/plugin-jsx',
                    '@react-vector-graphics/plugin-assets',
                ],
            },
        })
    })
})
