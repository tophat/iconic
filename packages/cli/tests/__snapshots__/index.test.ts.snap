// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli requests for missing option: output 1`] = `"Process exited with code: 1"`;

exports[`cli requests for missing option: pattern 1`] = `"Process exited with code: 1"`;
